### PR TITLE
feat(diagnostics): Detect duplicate overloads of functions and events

### PR DESCRIPTION
**Pull request by ggiraldez**
Builds on top of #2046

Closes NomicFoundation/slang-diagnostics-research#765
Closes NomicFoundation/slang-diagnostics-research#1290

### CODE REVIEW

File: crates/solidity-v2/testing/snapshots/binder_output/expressions/funcalls_ambiguous_overload/generated/0.8.0-failure.txt
Comment (teofr) [inline crates/solidity-v2/testing/snapshots/binder_output/expressions/funcalls_ambiguous_overload/generated/0.8.0-failure.txt:56]: I know it's unrelated to this PR, but should this be unresolved? Or ambiguous?

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/via_import_alias/generated/slang/0.8.0-failure.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# This file is generated automatically by infrastructure scripts. Please don't edit by hand.
+
+Diagnostics: 1
+
+[resolution/duplicate-function-definition] Error: Function with same name and parameter types defined twice.
```
Comment (teofr) [inline crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/via_import_alias/generated/slang/0.8.0-failure.txt:5]: Not too concerned about it, since solc's message is also quite confusing, but ideally the diagnostic should point to the imports that cause the error. Do you think this is achievable, maybe on a follow up PR?

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/qualified_import/input.sol [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+// --- path: consumer.sol
+// SPDX-License-Identifier: MIT
+pragma solidity *;
+
+// A qualified import binds one name to the whole file, so the declarations it
+// holds never join the importing file's overload sets — not even the two
+// indistinguishable ones below.
+import "./declarations.sol" as ns;
+import {f as aliased} from "./declarations.sol";
```
Comment (teofr) [inline crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/qualified_import/input.sol]: I'm not sure I understand this test, I feel there should be an `f` and a `g` declared in `consumer.sol` that would conflict with the ones at `declarations.sol`

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/library_and_interface/input.sol [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+library L {
+    function duplicated(uint256 a) internal pure returns (uint256) {
+        return a;
+    }
+
+    function duplicated(uint256 a) internal pure returns (uint256) {
+        return a + 1;
+    }
+
+    function ok(bytes32 a) internal pure returns (bytes32) {
```
Comment (teofr) [inline crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/library_and_interface/input.sol]: What's the point of the `ok` functions here? They're not really testing the overload mechanism since they have a different name.

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/contract_members/input.sol [hunk 4]
```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: MIT
```
Comment (teofr) [inline crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/contract_members/input.sol:1]: I feel  there's some tests missing from the function set:
- A three way duplicate check (https://github.com/argotorg/solidity/blob/v0.8.0/test/libsolidity/syntaxTests/freeFunctions/free_identical_multiple.sol?plain=1#L5), solc seems to diagnose the first appearance, we diagnose the last one, so this test is an interesting comparison
- A test like `via_import_alias` but where at the leaf file every function is aliased (and conflict), since we consider that point a new definition, I think it'd be interesting.
- A circular test https://github.com/argotorg/solidity/blob/v0.8.0/test/libsolidity/syntaxTests/multiSource/circular_import_3.sol?plain=1#L12, it should only fail with the last importing file, and it should fail for all three functions (I think)
- An inheritance case (solc seems to allow it, but it also seems to be a useless extra function definition, I don't think you can call the second bar...)
```solidity
// SPDX-License-Identifier: MIT
pragma solidity *;

contract A {}
contract B is A {}

function bar(A x)  {}
function bar(B x)  {}
```

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_event_definition/inherited/input.sol [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+// Two bases can also clash with each other, in a contract declaring neither.
+contract LeftParent {
+    event FromParents(uint256 a, uint256 b);
+}
+
+contract RightParent {
+    event FromParents(uint256 a, uint256 indexed b);
+}
+
+contract Diamond is LeftParent, RightParent {}
```
Comment (teofr) [inline crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_event_definition/inherited/input.sol]: I wouldn't call this a diamond, rather a join. But a diamond is an interesting test case:

```
contract Base {
    event FromParents(uint256 a, uint256 b);
}

contract LeftParent is Base {}

contract RightParent is Base {}

contract Diamond is LeftParent, RightParent {}
```

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_event_definition/contract_members/input.sol [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: MIT
```
Comment (teofr) [inline crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_event_definition/contract_members/input.sol:1]: I think we should add a test across files: https://github.com/argotorg/solidity/blob/v0.8.22/test/libsolidity/syntaxTests/events/file_level_event_duplicate_definition_via_import.sol?plain=1#L8

File: crates/solidity-v2/outputs/cargo/semantic/src/types/registry.rs [hunk 7]
```diff
@@ -712,6 +712,29 @@ impl TypeRegistry {
...
+    // Returns true if two parameter lists are indistinguishable once normalised
+    // for an external call, ie. the two declarations they belong to could not be
+    // told apart by a caller naming either of them.
+    //
+    // Normalisation only reaches the data location: `memory` and `calldata` are
+    // interchangeable because the ABI encodes both the same way, while
+    // `storage` stays distinct (a storage pointer is a different kind of
+    // argument altogether). This is the same relation as the one relaxing
+    // external overrides, applied to both lists in either direction.
+    pub(crate) fn parameter_lists_are_indistinguishable(
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/types/registry.rs:745]: nit: Not necessarily my suggestion, but maybe make it explicit that it's for external calls only.

```suggestion
    pub(crate) fn parameter_lists_are_indistinguishable_in_external_calls(
```

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/mod.rs [hunk 8]
```diff
@@ -196,6 +262,17 @@ impl<'a> HierarchyChecker<'a> {
     }
 }
 
+/// The members a base declares, whichever kind of type it is.
+fn members_of(binder: &Binder, base_id: NodeId) -> &[ir::ContractMember] {
+    match binder.find_definition_by_id(base_id) {
+        Some(Definition::Contract(contract)) => &contract.ir_node.members[..],
+        Some(Definition::Interface(interface)) => &interface.ir_node.members[..],
+        // A library is only ever the (sole) head of its own hierarchy.
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/mod.rs]: Why does this matter here?

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+            unreachable!("a single declaration was promoted just above");
+        };
+        let Some(parameter_types) = parameter_types_of(binder, types, definition) else {
+            return false;
+        };
+        if clashes_with_any(types, signatures, &parameter_types) {
+            return true;
+        }
+        signatures.push(parameter_types);
+        false
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: Could you shortly document the meaning of the return of this function?

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+    }
+}
+
+impl<'a> HierarchyChecker<'a> {
+    /// Reports this type's own functions that duplicate one declared before
+    /// them. Only the declarations written here take part, so this runs for the
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: "written here take part" is a bit confusing, took me a few reads to realized "here" means the contract (or interface or library) being analyzed, what do you think of


```suggestion
    /// them. Only the declarations written at the type being checked take part, so this runs for the
```

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+                    slot.insert(Overloads::Single(definition));
+                }
+                // A clash inside a shared base is revisited once per contract
+                // deriving from it, so each declaration is reported only once.
+                // The diagnostic is fully determined by that declaration, so
+                // the result doesn't depend on the order contracts are visited
+                // in.
+                Entry::Occupied(mut slot) => {
+                    if slot.get_mut().record(binder, types, definition)
+                        && reported.insert(definition.node_id())
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: I may have missed something, but we seem to check each definition once, what is the `reported` set guarding against?

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+        types: &TypeRegistry,
+        definition: &'a Definition,
+    ) -> bool {
+        // The second declaration is what forces the first one's signature.
+        // A declaration whose parameters aren't all typed yields none and takes
+        // no part in the comparison, exactly as if it were never recorded, so
+        // extending by the `Option` leaves the slot empty in that case.
+        if let Self::Single(first) = *self {
+            let mut signatures = Signatures::new();
+            signatures.extend(parameter_types_of(binder, types, first));
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:80]: Just to understand, the reason for `Overloads::Single` is mostly to postpone calculating the signature if only one element with a given name is present, right? If that's so, I'd get the parameter types of `definition` before moving it into `Several`

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 11 above)
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:159]: I may have missed something, but it seems we only check each definition once, is the check against `reported` guarding against anything?

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+/// Locates a declaration, for ordering an overload set spanning several files.
+fn declaration_site<'a>(
+    binder: &Binder,
+    file_node_mapper: &'a FileNodeMapper,
+    definition_id: NodeId,
+) -> (&'a FileId, usize) {
+    let start = binder
+        .find_definition_by_id(definition_id)
+        .map_or(0, |definition| definition.identifier().range.start);
+    (file_node_mapper.file_id_from_node_id(definition_id), start)
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:300]: This assumes that file ids follow the topological order of imports, but I'm not sure that's true (and it may definitely stop being true with parallelization). For example, if A.sol and B.sol both define a clashing function foo, and B imports A, but B is added to the compilation unit first (which is also the usual case today) its file id may be lower, therefore blaming the definition on A.

I think this test case could check it:


```solidity
// --- path: B.sol
// SPDX-License-Identifier: MIT
pragma solidity *;

import "./A.sol"'

function duplicated(uint256 a) pure returns (uint256) {
    return a;
}

// --- path: A.sol
// SPDX-License-Identifier: MIT
pragma solidity *;

function duplicated(uint256 b) pure returns (uint256) {
    return b + 1;
}
```

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+                .iter()
+                .filter_map(|definition_id| {
+                    let definition = binder
+                        .find_definition_by_id(*definition_id)
+                        .expect("visible declaration is registered");
+                    Some((definition, parameter_types_of(binder, types, definition)?))
+                })
+                .collect();
+
+            if declarations.len() < 2 {
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:213]: I would suggest doing the sorting after this then, I think it should be possible.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+    binder: &'a Binder,
+    file_scope: &'a FileScope,
+    visible: &mut Map<&'a str, Vec<NodeId>>,
+) {
+    // The closure is only precomputed for files that do have unqualified
+    // imports; without any, a file sees just its own declarations.
+    if file_scope.default_import_closure.is_empty() {
+        record_declarations(binder, file_scope, visible);
+        return;
+    }
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: This seems quite error prone, since it is also not computed early on on the passes. I'd suggest either always populating `default_import_closure` (only with itself when there's not default umports) or adding an accessor method that could do that behind the scenes (ie check if it's empty, if it is debug_assert that it has no default imports and return a vec with itself, etc)

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+}
+
+/// Records the functions and events `file_scope` declares, following import
+/// aliases so that the same declaration reached under one name through several
+/// paths is recorded once rather than clashing with itself.
+fn record_declarations<'a>(
+    binder: &'a Binder,
+    file_scope: &'a FileScope,
+    visible: &mut Map<&'a str, Vec<NodeId>>,
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: This surely could be a SmallVec


```suggestion
    visible: &mut Map<&'a str, SmallVec<[NodeId; 1]>>,
```

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+        for definition_id in definition_ids {
+            for target_id in underlying_declarations(binder, *definition_id) {
+                let Some(target) = binder.find_definition_by_id(target_id) else {
+                    continue;
+                };
+                if !matches!(target, Definition::Function(_) | Definition::Event(_)) {
+                    continue;
+                }
+                let declarations = visible.entry(symbol.as_str()).or_default();
+                if !declarations.contains(&target_id) {
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:283]: symbol doesnt change in these loops


```suggestion
        let declarations = visible.entry(symbol.as_str()).or_default();
        for definition_id in definition_ids {
            for target_id in underlying_declarations(binder, *definition_id) {
                let Some(target) = binder.find_definition_by_id(target_id) else {
                    continue;
                };
                if !matches!(target, Definition::Function(_) | Definition::Event(_)) {
                    continue;
                }
                if !declarations.contains(&target_id) {
```

**Comment by teofr** [comment]
Did a first pass, it looks good and it's mostly about some corner cases that I think would be worth to test

File: crates/solidity-v2/testing/snapshots/binder_output/expressions/funcalls_ambiguous_overload/generated/0.8.0-failure.txt
Comment (ggiraldez) [reply crates/solidity-v2/testing/snapshots/binder_output/expressions/funcalls_ambiguous_overload/generated/0.8.0-failure.txt:56]: Since the function types are equivalent (not identical because the associated definition is different) the second function is filtered in `filter_overriden_definitions`, which removes the ambiguity. This is the expected behaviour if each declaration was in a different contract in the hierarchy.

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/via_import_alias/generated/slang/0.8.0-failure.txt (hunk 1 above)
Comment (ggiraldez) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/via_import_alias/generated/slang/0.8.0-failure.txt:5]: Agree this can be very confusing. Pointing to the import should be possible, but I think it would still be incomplete information. To be able to report this accurately we should be able to provide a stack of locations which is something our diagnostics don't support right now. I imagine something like:

```
[resolution/duplicate-function-definition] Error: Function with same name and parameter types defined twice.
   ╭─[declarations.sol:8:1]
   │
 8 │ function g() pure returns (uint256) {
   │ ──────────────────┬─────────────────  
   │                   ╰─────────────────── Error occurred here.
───╯
... while compiling 'consumer.sol' as symbol 'f', visible through
   ╭─[consumer.sol:4:1]
   │
 4 │ import "./alias.sol";
   │ ───────┬────────────  
   │        ╰─────────────────── Import location
───╯
   ╭─[alias.sol:4:1]
   │
 4 │ import {f as g} from "./declarations.sol";
   │ ───────┬────────────  
   │        ╰─────────────────── Import location
───╯

... conflicts with
   ╭─[declarations.sol:8:1]
   │
 4 │ function f() pure returns (uint256) {
   │ ──────────────────┬─────────────────  
   │                   ╰─────────────────── Previous declaration.
───╯
... visible through
   ╭─[consumer.sol:6:1]
   │
 6 │ import "./declarations.sol";
   │ ───────┬────────────  
   │        ╰─────────────────── Import location
───╯
```

Achievable, but not with the current infrastructure. This example is particularly confusing, but there are many other instances where having diagnostic stacks would be helpful.

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/qualified_import/input.sol (hunk 2 above)
Comment (ggiraldez) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/qualified_import/input.sol]: Yeah, you're right. I think this is trying to show that if they are under a different namespace, they don't interfere. But it's not clear. I'll improve it.

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/library_and_interface/input.sol (hunk 3 above)
Comment (ggiraldez) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/library_and_interface/input.sol]: You're right. I'll change them to valid overloads of the `duplicated` function.

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/contract_members/input.sol (hunk 4 above)
Comment (ggiraldez) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/contract_members/input.sol:1]: > * A three way duplicate check (https://github.com/argotorg/solidity/blob/v0.8.0/test/libsolidity/syntaxTests/freeFunctions/free_identical_multiple.sol?plain=1#L5), solc seems to diagnose the first appearance, we diagnose the last one, so this test is an interesting comparison

Adding. We diagnose every duplicate after the first declaration.

> * A test like `via_import_alias` but where at the leaf file every function is aliased (and conflict), since we consider that point a new definition, I think it'd be interesting.

I don't quite understand what you'd be trying to test here. Different functions, all aliased to the same name in the leaf file? Or the same function, brought in by multiple aliases, but then joined in a single symbol?

> * A circular test https://github.com/argotorg/solidity/blob/v0.8.0/test/libsolidity/syntaxTests/multiSource/circular_import_3.sol?plain=1#L12, it should only fail with the last importing file, and it should fail for all three functions (I think)

Adding. Only two diagnostics (for the functions in s2) are emitted because we're tracking the underlying functions to avoid duplicate reports.

> * An inheritance case (solc seems to allow it, but it also seems to be a useless extra function definition, I don't think you can call the second bar...)
> 
> ```solidity
> // SPDX-License-Identifier: MIT
> pragma solidity *;
> 
> contract A {}
> contract B is A {}
> 
> function bar(A x)  {}
> function bar(B x)  {}
> ```

This is not a duplicate definition validation. The types `A` and `B` are distinct. It does show an ambiguous resolution diagnostic when you try to invoke it, but that's a different thing, maybe worth adding in #2056.

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_event_definition/inherited/input.sol (hunk 5 above)
Comment (ggiraldez) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_event_definition/inherited/input.sol]: True it's not a diamond. I'll rename to `Join`. You're example doesn't trigger though, because the declaration is in the base and it's tracked there.

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_event_definition/contract_members/input.sol (hunk 6 above)
Comment (ggiraldez) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_event_definition/contract_members/input.sol:1]: Added.

File: crates/solidity-v2/outputs/cargo/semantic/src/types/registry.rs (hunk 7 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/types/registry.rs:745]: I'll push back here because even though the normalisation is equivalent to what's done for external calls, this indistiguishability algorithm is applied regardless of the visibility of the function, even internal vs. external functions. Eg.

```solidity
contract C {
  function foo(bytes memory b) internal pure {}
  function foo(bytes calldata b) external pure {}
}
```

Produces the error even though the functions are invoked in completely different conditions.

I updated the function description to better reflect this.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/mod.rs (hunk 8 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/mod.rs]: It doesn't. Probably a stale comment after the many perf mitigation refactors this algorithm went through. I removed it.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 9 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: I updated the function description to make it a bit more explicit and added a comment in the end.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 10 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: Agreed. There's a parameter `declared_here` in `redeclarations.rs` with explicit documentation, but "here" is not meaningful. I think I remember there were more similar instances. I'll try to come up with a better framing.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 12 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:80]: Fair point, although it shouldn't make much difference in practice. We'd fail to get the parameter types only on an invalid program.

But to answer your question, yes, that's the purpose. The majority of the functions should not be overloaded, and computing the parameter types is costly. This avoids doing unnecessary work.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 11 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:159]: Yes, but base contracts may be folded multiple times so this avoid the duplicate diagnostics. Eg.
```solidity
contract Base {
  event E();
  event E();
}
contract Foo is Base {}
contract Bar is Base {}
```
would report twice for the duplication in `Base`.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 13 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:300]: This does use order of source ingestion, but right now it doesn't make any guarantees wrt import order. The stability of the order is important to be consistent as to the diagnostic locations, but nothing else. Since we don't have stack frames for diagnostics, I don't think it matters too much on which location we report the error.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 14 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:213]: Again, shouldn't make much difference since we have a similar guard before sorting. This is in case typing info is incomplete and we lost some declarations in the way, in which case the program is not valid.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 15 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: I saw this as well, but it predates this PR. `resolve_in_file_scope` uses this fact to shortcut resolution and avoid allocation of an intermediate vector of definitions. It does use `file_scope.default_imports` though so I think it makes sense we do the same here. I'll change it.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 16 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: Yes, good point.

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 17 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:283]: This change is not good. It allocates an entry in `visible` for every symbol, regardless of whether it needs to be checked or not (ie. even when it's neither a function nor an event).

File: crates/solidity-v2/testing/snapshots/binder_output/expressions/funcalls_ambiguous_overload/generated/0.8.0-failure.txt
Comment (teofr) [reply crates/solidity-v2/testing/snapshots/binder_output/expressions/funcalls_ambiguous_overload/generated/0.8.0-failure.txt:56]: Thanks!

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/via_import_alias/generated/slang/0.8.0-failure.txt (hunk 1 above)
Comment (teofr) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/via_import_alias/generated/slang/0.8.0-failure.txt:5]: Makes sense, thanks!

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/contract_members/input.sol (hunk 4 above)
Comment (teofr) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/contract_members/input.sol:1]: > I don't quite understand what you'd be trying to test here. Different functions, all aliased to the same name in the leaf file? Or the same function, brought in by multiple aliases, but then joined in a single symbol?

Since we treat import alias as definitions, I thought it could make sense to test that diagnostics will follow the symbols to the original definition, rather than diagnose on the alias itself, after looking at the rest of the PR, I don't think it's as needed anymore, so feel free to ignore it.

> This is not a duplicate definition validation. The types A and B are distinct. It does show an ambiguous resolution diagnostic when you try to invoke it, but that's a different thing, maybe worth adding in https://github.com/NomicFoundation/slang/pull/2056.

I guess it would test it doesn't trigger, but yeah, maybe it belongs in that PR.

Thanks!

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/qualified_import/input.sol (hunk 2 above)
Comment (teofr) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/qualified_import/input.sol]: Thanks!

File: crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/library_and_interface/input.sol (hunk 3 above)
Comment (teofr) [reply crates/solidity-v2/testing/snapshots/diagnostics_output/resolution/duplicate_function_definition/library_and_interface/input.sol]: Thanks!

File: crates/solidity-v2/outputs/cargo/semantic/src/types/registry.rs (hunk 7 above)
Comment (teofr) [reply crates/solidity-v2/outputs/cargo/semantic/src/types/registry.rs:745]: Makes sense, thank you!

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 9 above)
Comment (teofr) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: Thank you!

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 12 above)
Comment (teofr) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:80]: Makes sense. I didn't realize that `parameter_types_of` should not fail on correct programs, I missed the early filter on `Function` and `Event`, thank you!

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 11 above)
Comment (teofr) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:159]: Makes sense, missed that `reported_duplicate` is shared throughout p4

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 13 above)
Comment (teofr) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:300]: Oh, ok, I made assumptions on those guarantees. Thanks!

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 15 above)
Comment (teofr) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs]: Thanks!

File: crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs (hunk 17 above)
Comment (teofr) [reply crates/solidity-v2/outputs/cargo/semantic/src/passes/p4_compute_linearisations/duplicate_declarations.rs:283]: You're right! I still think is possible with `LazyCell`, or by creating the SmallVec and only inserting if it's not empty, but probably not a big win for a more complicated function.

**Review by teofr — APPROVED** [review]
Thanks for the patience with the many comments, looks good on my side!
